src/sync/peer-sync-controller.js

```diff
@@ -0,0 +1,278 @@
+
+    // Always need to replicate the project creator core
+    coreManager.creatorCore.replicate(protomux)
+    this.#replicatingCores.add(coreManager.creatorCore)
+
+    coreManager.on('add-core', this.#handleAddCore)
+    syncState.on('state', this.#handleStateChange)
+
+    this.#updateEnabledNamespaces()
+  }
+
+  /**
+   * Enable syncing of data (in the data and blob namespaces)
+   */
+  enableDataSync() {
+    this.#isDataSyncEnabled = true
+    this.#updateEnabledNamespaces()
+  }
+
+  /**
+   * Disable syncing of data (in the data and blob namespaces).
+   *
+   * Syncing of metadata (auth, config and blobIndex namespaces) will continue
+   * in the background without user interaction.
+   */
+  disableDataSync() {
+    this.#isDataSyncEnabled = false
+    this.#updateEnabledNamespaces()
+  }
+
+  /**
+   * Handler for 'core-add' event from CoreManager
+   * Bound to `this` (defined as static property)
+   *
+   * @param {import("../core-manager/core-index.js").CoreRecord} coreRecord
+   */
+  #handleAddCore = ({ core, namespace }) => {
+    if (!this.#enabledNamespaces.has(namespace)) return
+    this.#replicateCore(core)
+  }
+
+  /**
+   * Handler for 'state' event from SyncState
+   * Bound to `this` (defined as static property)
+   *
+   * @param {import("./sync-state.js").State} state
+   */
+  #handleStateChange = async (state) => {
+    // The remotePublicKey is only available after the noise stream has
+    // connected. We shouldn't get a state change before the noise stream has
+    // connected, but if we do we can ignore it because we won't have any useful
+    // information until it connects.
+    if (!this.#protomux.stream.remotePublicKey) return
+    const peerId = this.#protomux.stream.remotePublicKey.toString('hex')
+    this.#syncStatus = getSyncStatus(peerId, state)
+    const localState = mapObject(state, (ns, nsState) => {
+      return [ns, nsState.localState]
+    })
+
+    // Map of which namespaces have received new data since last sync change
+    const didUpdate = mapObject(state, (ns) => {
+      const nsDidSync =
+        this.#prevSyncStatus[ns] !== 'synced' &&
+        this.#syncStatus[ns] === 'synced'
+      const prevNsState = this.#prevLocalState[ns]
+      const nsDidUpdate =
+        nsDidSync &&
+        (prevNsState === null || prevNsState.have !== localState[ns].have)
+      if (nsDidUpdate) {
+        this.#prevLocalState[ns] = localState[ns]
+      }
+      return [ns, nsDidUpdate]
+    })
+    this.#prevSyncStatus = this.#syncStatus
+
+    if (didUpdate.auth) {
+      try {
+        const cap = await this.#capabilities.getCapabilities(peerId)
+        this.#syncCapability = cap.sync
+      } catch (e) {
+        // Any error, consider sync blocked
+        this.#syncCapability = createNamespaceMap('blocked')
+      }
+    }
+    // console.log(peerId.slice(0, 7), this.#syncCapability)
+    // console.log(peerId.slice(0, 7), didUpdate)
+    // console.dir(state, { depth: null, colors: true })
```

Comment:
```suggestion
```

Reply:
I'll keep these for now for debugging integration tests.